src/sound.c

```diff
@@ -199,4 +205,151 @@ sound_free(void)
 }
 #endif
 
-#endif  // FEAT_SOUND && HAVE_CANBERRA
+#elif defined(MSWIN) // MSWIN
+
+static HWND g_hWndSound = NULL;
+
+    LRESULT CALLBACK
+sound_wndproc(HWND hwnd, UINT message, WPARAM wParam, LPARAM lParam)
+{
+    soundcb_T	*p;
+
+    switch(message)
+    {
+	case MM_MCINOTIFY:
+	    for (p = first_callback; p != NULL; p = p->snd_next)
+		if (p->device_id == (MCIDEVICEID) lParam)
+		{
+		    typval_T	argv[3];
+		    typval_T	rettv;
+		    int		dummy;
+
+		    argv[0].v_type = VAR_NUMBER;
+		    argv[0].vval.v_number = p->sound_id;
+		    argv[1].v_type = VAR_NUMBER;
+		    argv[1].vval.v_number = wParam == MCI_NOTIFY_SUCCESSFUL ? 0
+					  : wParam == MCI_NOTIFY_ABORTED ? 1 : 2;
+		    argv[2].v_type = VAR_UNKNOWN;
+
+		    call_callback(&p->snd_callback, -1,
+					    &rettv, 2, argv, NULL, 0L, 0L, &dummy, TRUE, NULL);
+		    clear_tv(&rettv);
+
+		    delete_sound_callback(p);
+		    redraw_after_callback(TRUE);
+
+		}
+	    break;
+    }
+
+    return DefWindowProc(hwnd, message, wParam, lParam);
+}
+
+    static HWND
+sound_window()
+{
+    if (g_hWndSound == NULL)
+    {
+	LPCTSTR clazz = "VimSound";
+	WNDCLASS wndclass = { 0, sound_wndproc, 0, 0, g_hinst, NULL, 0, 0, NULL, clazz };
+	RegisterClass(&wndclass);
+	g_hWndSound = CreateWindow(clazz, NULL, 0, 0, 0, 0, 0,
+		HWND_MESSAGE, NULL, g_hinst, NULL);
+    }
+
+    return g_hWndSound;
+}
+
+    void
+f_sound_playevent(typval_T *argvars, typval_T *rettv)
+{
+    smsg("sound_playevent() not implemented");
+}
+
+    void
+f_sound_playfile(typval_T *argvars, typval_T *rettv)
+{
+    int		newid = sound_id + 1;
+    int		len;
+    char_u	*p, *esc;
+    soundcb_T	*soundcb;
+
+    rettv->v_type = VAR_NUMBER;
+
+    esc = vim_strsave_shellescape(
+		   tv_get_string(&argvars[0]), FALSE, FALSE);
+
+    len = STRLEN(esc) + 5 + 18 + 1;
+    p = alloc(len);
+    vim_snprintf((char *)p, len, "open %s alias sound%06d", esc, newid);
+    free(esc);
+    if (mciSendString((char*) p, NULL, 0, (HWND) sound_window()) != 0)
```

Comment:
Don't we need to convert the encoding of the sound filename to UTF-16 and use the wide function?
Sound files with multibyte filename might not work with the current implementation.

Reply:
Yes, will do. I will have to use wide string APIs and window proceudure too.